wcstools-3.9.2/man/man1/simpos.1
.TH simpos(1) WCSTools "14 May 2015"
.SH Name
simpos \- Find position of a named object in the sky
.SH Synopsis
simpos [options] [object name]
.SH Description
.I simpos
is a utility for finding the coordinates of any named object in the sky
using the SIMBAD name resolver at either the CDS in Strasbourg or the
ADS in Cambridge, MA.  It can search SIMBAD, Vizier, and/or NED (NASA
Extragalactic Database) in any order.
.SH Options
.TP
.B object name
Name of object for which to find coordinates.
.TP
.B \@listfile
File with one object name per line.
.TP
.B \-b <RA> <Dec>
Output B1950 (FK4) coordinates instead of default J2000 (FK5)
.TP
.B \-c
Use the CDS server in Strasbourg, France rather than the default server
at the Astrophysical Data System in Cambridge, Massachusetts, USA.
.TP
.B \-d
 Output RA and Dec in degrees instead of hh:mm:ss dd:mm:ss
.TP
.B \-e
Output ecliptic coordinates
.TP
.B \-g
Output galactic coordinates
.TP
.B \-i
Print object ID(s) returned
.TP
.B \-n
Find coordinates using NED
.TP
.B \-s 
Find coordinates using SIMBAD
.TP
.B \-t
Print output as tab-separated table
.TP
.B \-v
Verbose listing of results of search
.TP
.B \-z
Find coordinates in the Vizier catalogs

.SH See Also
scat()

.SH Web Page
http://tdc-www.harvard.edu/software/wcstools/scat/

.SH Author
Doug Mink, SAO (dmink@cfa.harvard.edu)


wcstools-3.9.2/NEWS
Version 3.9.2 (May 15, 2015)
imcat, imstar: Print two decimal places on PROS region file coordinates
simpos: Switch to CfA Vizier server as default (2015-05-05)
simpos: Use NED name resolver if called as nedpos or using -n argument
simpos: Look up in VIZIER catalogs if called as vizpos or using -z argument
simpos: Use CDS server if called as cdspos or using -c argument
nedpos,nedname: Drop tasks because IPAC has dropped interface; use simpos -n

Version 3.9.1 (March 24, 2015)
remap: Declare GetFITSWCS() as well as GetWCSFITS()
cphead: Fix bug freeing header string more than once
imcat,scat,immatch,imwcs: Fix bugs with web access to GSC2 and SDSS catalogs
gsc2read.c: Drop concatenation of "blank" string in middle of search URL
webread.c: Fix bug in URL filtering

Version 3.9.0 (December 19, 2014)
gethead: Add -r option to return multiple values as a single word separated
         by the first character of the keyword the value of which follows
         (2014-12-19)
remap: Update usage to -f WCSfile suggested by Steve Willner (2014-02-19)
sethead: Fix bug dealing with setting the same keyword in multiple file extensions
imcat, scat, immatch, imwcs: Implement UCAC4 catalog
fileutil.c: Add next_line() to return one line of file
fitfile.c: fix buffer reallocation bug in fitsrhead()
tabread.c: fix file tests
gsc2read.c and tabread.c: adjust to GSC2.3.3 format
webbuff.c: Add filter to convert characters in web URLs

Version 3.8.9 (September 16, 2013)
imcat, imwcs, immatch, scat: Work around SDSS server failure
sdssread.c: Add alternate UK and US catalog servers

Version 3.8.8 (April 30, 2013)
getdate: Add -s option to subtract (or add) days
getdate: Add self-conversion for *JD's
getfits: Print error, not blank file, if requested region is off image
simpos: Update URL for SIMBAD query

Version 3.8.7 (October 31, 2012)
Rename Man to man to avoid problems with Mac OSX file name convention
filedir: If pathname ends in "/", drop last directory
catutil.c:  Skip trailing right bracket in aget*()
dateutil.c:  Unused l0 dropped from jd2lst(); ts2ss from jd2mst()
imio.c:  Fix errors with short and character images in minvec(), maxvec()
wcs.c:  Drop d1 and d2 from wcsdist(); diffi from wcsdist1()
wcs.c:  Drop depwcs; it's in main wcs structure
wcsinit.c: Drop unused variable iszpx; fix bug in latmin assignment
zpxpos.c: Fix code for quadratic near pole

Version 3.8.6 (August 10, 2012)
sumpix: Add options i to print index and e to compute medians
gethead: Call fk425e() instead of fk425()
imcat: Drop declaration of unused fk425e()
imutil.c: Add medpix() and meanpix() to generalize across pixel size
imutil.c: Add medpixi1() and meanpixi1() to handle 8-bit images
imio.c:  Fix 8-bit variables to be unsigned char
ctgread.c: Allow up to 9 magnitudes (/9) instead of 4
catutil.c, polfit.c: Fix xterm computation in polcomp() from + to *

Version 3.8.5 (April 12, 2012)
getdate: Fix string overflow in ts0 assignment (2012-03-27)
gethead: If length of header/file content string is zero, exit with error
sumpix: Improve argument parsing so order doesn't matter (2012-02-01)
matchstar.c: Fix bug in WCSMatch() which miscomputed means
fitsfile.c: Always check first 8 characters of FITS files for "SIMPLE"
imhfile.c: Fix potential overflow in newpixfile assignment (2012-03-27)
imio.c: Change 8-bit pixel values from char to unsigned char

Version 3.8.4 (September 1, 2011)
cphead, keyhead, sethead: Fix overflow bug by increasing size of history from 72 to 128
imgetwcs.c, wcsinit.c, wcs.c, wcs.h, worldpos.c: Add TPV WCS for TAN with PV terms

Version 3.8.3 (August 17, 2011)
getdate: Print hjd in help message instead of jhd
gethead: First column name is FILE_NAME if FILENAME keyword is used (2011-06-29)
gethead: Fix memory leaks (2011-05-19)
gethead: Add -j option to convert RA and Dec to J2000 (2011-05-20)
gethead: Fix -x option for named or unnamed range of extensions (2011-08-12)
gettab: Fix variable name bug so program works!
scat: Fix tab spacing in tmce(2MASS catalog with magnitude errors) (2011-04-21) 
imcat,scat,imwcs,immatch: Fix problem with UCAC3 (2011-04-27)
skycoor: Deal with comma in command line input coordinates (2011-04-19)
skycoor: Drop 99 for skipped RA in input list file (2011-08-17)
hget.c: Free allocated memory in strnsrch() to eliminate memory leak
imhfile.c: Free *newpixname* not pixname. (2011-05-20)
ned_cli.c: Declard *d double, not long double, in ned_getdouble() (2011-05-20)
ucacread.c: Find stars when RA is unevenly distributed in a declination zone
ucacread.c: Drop extra formats from printf() (2011-05-20)
wcs.c: Fix wcsfree() so it frees depended-on WCS structures (2011-05-09)
wcsinit.c: Change error() calls to setwcserr()
wcslib.h: Declare undeclared SCAMP subroutine raw-to-pv()

Version 3.8.2 (March 18, 2011)
all: Add support for NOAO ZPX projection with code from Frank Valdes
sethead: Exit with error errflag
fixpix: Increase maximum number of regions from 10 to 100
fixpix: Log number of regions fixed
remap: Keep center sky coordinates if not set on command line
getpix: Fix mean and limits options
getpix: Add -t option for tab separated tables
fixpix: If filling from pixel list file, put file name in header
tmcread.c: Fix 2MASS catalog search bug by adding last star to search
zpxpos.c, wcs.c, wcsinit.c: Add support for NOAO ZPX protection (Frank Valdes)
proj.c: If no PV coefficients in ZPN projection, make it ARC
proj.c: Add code to project SCAMP polynomical TAN
wcsinit.c: Add code to read output from SCAMP (fixed again on 2011-03-18)
ucacread.c: Fix bug reading UCAC3 catalog
wcs.c: Fix bug in wcsc2pix() when using WCSDEP

Version 3.8.1 (May 27, 2010)
scat,imcat: Set GSC2 magnitudes > 90 to 99.99
imwcs, immatch: Fix for catalogs with lots of magnitudes
fixpix: Add -c and -r options to force interpolation direction
gethead: Fix buffer reallocation bug which crashed reading very large headers
gethead: Fix trailing spaces on ASCII file quoted string values

wcstools-3.9.2/simpos.c
/*** simpos.c - search object by its name from command line arguments
 *** May 15, 2015
 *** By Jessica Mink, sort of after IPAC byname.c for searching NED
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "libwcs/wcs.h"
#include "libwcs/fitsfile.h"
#include "libwcs/wcscat.h"

extern int   ned_errno;
static char searchorder[4];
static int printall = 0;
static void PrintUsage();
static char *RevMsg = "SIMPOS 3.9.2, 15 May 2015, Jessica Mink SAO";

int
main (ac, av)
int  ac;
char *av[];
{
   
    int lobj;
    int i;
    int verbose = 0;
    int printid = 0;
    int printname = 0;
    int printdeg = 0;
    int nid = 0;
    int nobj = 0;
    int nfobj = 0;
    int cds = 0;
    int ned = 0;
    int viz = 0;
    int simbad = 0;
    int outsys = WCS_J2000;
    int sysj = WCS_J2000;
    int tabout = 0;
    double ra, dec;
    char *str, *objname, *posdec, *posra, *iend, *ieq;
    char *listfile;
    char rastr[32], decstr[32];
    char newobj[32];
    char *buff, *buffid, *idline, *posline, *errline, *id, *errend;
    char url[256];
    char *xbuff;
    int lbuff;
    FILE *flist;
    char cr, lf, eq;

    listfile = NULL;
    cr = (char) 13;
    lf = (char) 10;
    eq = '=';
    searchorder[0] = (char) 0;
    searchorder[1] = (char) 0;
    searchorder[2] = (char) 0;
    searchorder[3] = (char) 0;

    if ( !strcmp (*av, "cdspos") )
	cds = 1;
    if ( !strcmp (*av, "nedpos") ) {
	searchorder[0]='N';
	ned = 1;
	}
    if ( !strcmp (*av, "vizpos") ) {
	searchorder[0]='V';
	viz = 1;
	}

    str = *(av+1);
    if (!str || !strcmp (str, "help") || !strcmp (str, "-help"))
	PrintUsage (NULL);
    if (!strcmp (str, "version") || !strcmp (str, "-version"))
	PrintUsage ("version");


    /* crack arguments */
    for (av++; --ac > 0 && (**av == '-' || **av == '@'); av++) {

	if (**av == '-') {
            char c;
	    str = *av;
            while (c = *++str) {
            switch (c) {

        	case 'a':       /* Print all ID information */
	            printall = 1;
		    break;

        	case 'b':       /* Print coordinates in B1950 */
	            outsys = WCS_B1950;
		    break;

        	case 'c':       /* Use CDS server in France */
	            cds = 1;
		    break;

		case 'd':       /* Print coordinates in degrees */
		    printdeg++;
		    break;

		case 'e':       /* Print coordinates in ecliptic coordinates */
		    outsys = WCS_ECLIPTIC;
		    break;

		case 'g':       /* Print coordinates in galactic coordinates */
		    outsys = WCS_GALACTIC;
		    break;

		case 'i':       /* Print all IDs found by SIMBAD */
		    printid++;
		    break;

		case 'n':       /* Search NED database */
		    ned++;
		    strcat (searchorder, "N");
		    break;

		case 'o':       /* Print object name */
		    printname++;
		    break;

		case 's':       /* Database search order */
		    simbad++;
		    strcat (searchorder, "S");
		    break;

		case 't':       /* Print output in tab-separated table */
		    tabout++;
		    break;

		case 'v':       /* More verbosity, including first ID */
		    verbose++;
		    break;

		case 'z':       /* Search Vizier catalogs */
		    viz++;
		    strcat (searchorder, "V");
		    break;

		default:
		    PrintUsage(NULL);
		    break;
		}
		}
	    }

	/* File containing a list of object names */
	else if (**av == '@') {
	    listfile = *av + 1;
	    if (isfile (listfile)) {
		nfobj = getfilelines (listfile);
		if (verbose)
		    fprintf (stderr,"SIMPOS: %d objects from file %s\n",
			     nfobj, listfile);
		if ((flist = fopen (listfile, "r")) == NULL) {
		    fprintf (stderr,"SIMPOS: List file %s cannot be read\n",
			     listfile);
		    }
		}
	    else {
		printf ("SIMPOS: List file %s does not exist\n", listfile);
		listfile = NULL;
		}
	    
	    }
	}

    /* There are ac remaining file names starting at av[0] */
    if (ac == 0 && !listfile)
        PrintUsage (NULL);

    /* Set searches to SNV if not otherwise set */
    if ( searchorder[0] == (char) 0 )
	strcpy (searchorder, "SNV");
    if (listfile) {
	ac = nfobj;
	objname = newobj;
	}
    while (ac > 0) {

	if (listfile)
	    fgets (newobj, 32, flist);
	else
	    objname = *av++;
	ac--;

	/* Replace underscores and spaces with plusses */
	lobj = strlen (objname);
	for (i = 0; i < lobj; i++) {
	    if (objname[i] == '_')
		objname[i] = '+';
	    if (objname[i] == ' ') {
		if (i == lobj-1)
		    objname = (char) 0;
		else
		    objname[i] = '+';
		}
	    }
	if (verbose)
	    printf ("%s -> ", objname);

	for (i = 0; i < 256; i++) {
	    url[i] = (char) 0;
	    }
	if (cds)
	    strcpy (url, "http://cdsweb.u-strasbg.fr/cgi-bin/nph-sesame/-oI/");
	else
	    strcpy (url, "http://vizier.cfa.harvard.edu/viz-bin/nph-sesame/-oI/");
	strcat (url, searchorder);
	if ((printid || printall) && !strcmp (searchorder, "V"))
	    strcat (url, "A");
	strcat (url, "?");
	strcat (url, objname);
	if (verbose)
	    printf ("%s\n", url);
	buff = webbuff (url, verbose, &lbuff);

	if (buff == NULL) {
	    if (verbose)
		printf ("no return from SIMBAD\n");
	    else {
		if (ned)
		    fprintf (stderr,"*** No return from NED for %s\n",objname);
		if (viz)
		    fprintf (stderr,"*** No return from Vizier for %s\n",objname);
		if (simbad)
		    fprintf (stderr,"*** No return from SIMBAD for %s\n",objname);
		}
	    continue;
	    }
	else if (printall) {
	    buff[lbuff] = (char) 0;
	    xbuff = buff + lbuff - 20;
	    (void) printf ("%s\n", xbuff);
	    (void) printf ("%s\n", buff);
	    (void) printf ("End of Buffer\n");
	    continue;
	    }

	/* Read number of objects identified */
	if ((idline = strsrch (buff, "#=")) != NULL) {
	    id = strchr (idline, ':');
	    if (id != NULL)
		nid = atoi (id+2);
	    else
		nid = 0;

	    /* Get position */
	    if ((posline = strsrch (buff, "%J ")) != NULL) {
		posra = posline + 3;
		while (*posra == ' ')
		    posra++;
		ra = atof (posra);
		posdec = strchr (posra, ' ');
		while (*posdec == ' ')
		    posdec++;
		posline = strchr (posdec, ' ');
		dec = atof (posdec);
		}
	    else {
		if (verbose)
		    printf ("no position from SIMBAD\n");
		else  {
		    if (ned)
		        fprintf (stderr,"*** No NED position for %s\n",objname);
		    if (viz)
		        fprintf (stderr,"*** No Vizier position for %s\n",objname);
		    if (simbad)
		        fprintf (stderr,"*** No SIMBAD position for %s\n",objname);
		    }
		continue;
		}
	    }

	else {
	    nid = 0;
	    if ((errline = strsrch (buff, "#!SIMBAD: ")) != NULL) {
		if ((errend = strchr (errline, '\n')) != NULL)
		   *errend = (char) 0;
		fprintf (stderr, "*** %s\n", errline+10);
		}
	    else {
		fprintf (stderr, "*** No SIMBAD position for %s\n", objname);
		}
	    continue;
	    }

	if (nid > 0) {
	    if (verbose) {
		if (ned && nid == 1)
		    fprintf (stdout, "%d object found in NED: \n", nid);
		else if (ned)
		    fprintf (stdout, "%d objects found in NED: \n", nid);
		else if (viz && nid == 1)
		    fprintf (stdout, "%d object found by VIZIER: \n", nid);
		else if (viz)
		    fprintf (stdout, "%d objects found by VIZIER: \n", nid);
		else if (nid == 1)
		    fprintf (stdout, "%d object found by SIMBAD: \n", nid);
		else
		    fprintf (stdout, "%d objects found by SIMBAD: \n", nid);
		}

	/* Print Starbase header if requested */
	    nobj++;
	    if (tabout && nobj == 1) {
		if (ned)
		    printf ("catalog	NED\n");
		else if (viz)
		    printf ("catalog	VIZIER\n");
		else
		    printf ("catalog	SIMBAD\n");
		if (outsys == WCS_GALACTIC)
		    printf ("radecsys	galactic\n");
		else if (outsys == WCS_ECLIPTIC)
		    printf ("radecsys	ecliptic\n");
		else if (outsys == WCS_B1950)
		    printf ("radecsys	B1950\n");
		else
		    printf ("radecsys	J2000\n");
		if (outsys == WCS_B1950) {
		    printf ("equinox	1950.0\n");
		    printf ("epoch	1950.0\n");
		    }
		else {
		    printf ("equinox	2000.0\n");
		    printf ("epoch	2000.0\n");
		    }
		printf ("program	%s\n",RevMsg);
		if (printname)
		    printf ("id    	");
		printf ("ra      	dec       ");
		printf ("\n");
		if (printname)
		    printf ("------	");
		printf ("------------	------------");
		printf ("\n");
		}
	    if (printname) {
		printf ("%s", objname);
		if (tabout)
		    printf ("	");
		else
		    printf (" ");
		}
	    if (outsys != WCS_J2000)
		wcscon (sysj, outsys, 0.0, 0.0, &ra, &dec, 0.0);
	    if (outsys == WCS_ECLIPTIC || outsys == WCS_GALACTIC) {
		if (verbose)
		    fprintf (stdout, "l= ");
		fprintf (stdout, "%.6f", ra);
		if (tabout)
		    printf ("	");
		else
		    printf (" ");
		if (verbose)
		    fprintf (stdout, "b= ");
		if (dec >= 0.0)
		    fprintf (stdout, "+");
		fprintf (stdout, "%.6f", dec);
		if (!tabout) {
		    if (outsys == WCS_GALACTIC)
			fprintf (stdout, " Galactic");
		    else if (outsys == WCS_ECLIPTIC)
			fprintf (stdout, " Ecliptic");
		    }
		fprintf (stdout, "\n");
		}
	    else {
		if (verbose)
		    fprintf (stdout, "ra= ");
		if (printdeg)
		    fprintf (stdout, "%.6f", ra);
		else {
		    ra2str (rastr, 31, ra, 3);
		    fprintf (stdout, "%s", rastr);
		    }
		if (tabout)
		    printf ("	");
		else
		    printf (" ");
		if (verbose)
		    fprintf (stdout, "dec=");
		if (printdeg)
		    fprintf (stdout, "%.6f", dec);
		else {
		    dec2str (decstr, 31, dec, 2);
		    fprintf (stdout, "%s", decstr);
		    }
		if (!tabout) {
		    if (outsys == WCS_B1950)
			fprintf (stdout, " B1950");
		    else
			fprintf (stdout, " J2000");
		    }
		fprintf (stdout, "\n");
		}
	    if (printid) {
		buffid = buff;
		while ((buffid = strsrch (buffid, "%I ")) != NULL) {
		    buffid = buffid + 3;
		    if ((iend = strchr (buffid, lf))) {
			*iend = (char) 0;
			if ((ieq = strchr (buffid, eq)))
			    *ieq = (char) 0;
			printf ("%s\n", buffid);
			if (ieq)
			    *ieq = eq;
			*iend = lf;
			}
		    else if ((iend = strchr (buffid,cr))) {
			*iend = (char) 0;
			if ((ieq = strchr (buffid, eq)))
			    *ieq = (char) 0;
			printf ("%s\n", buffid);
			if (ieq)
			    *ieq = eq;
			*iend = cr;
			}

		    }
		}
	    }
	free (buff);
	}
    exit (0);
}

static void
PrintUsage (command)

char	*command;	/* Command where error occurred or NULL */

{
    fprintf (stderr,"%s\n", RevMsg);
    fprintf (stderr,"Return RA and Dec for object name using SIMBAD, NED, Vizier\n");
    fprintf (stderr,"Usage:  simpos [-idtv][b|e|g] name1 name2 ...\n");
    fprintf (stderr,"        simpos [-idtv][b|e|g] @namelist ...\n");
    fprintf (stderr,"name(n): Objects for which to search (space -> _)\n");
    fprintf (stderr,"namelist: File with one object name per line\n");
    fprintf (stderr,"-a: Print all information returned\n");
    fprintf (stderr,"-b: Print coordinates in B1950 instead of J2000\n");
    fprintf (stderr,"-c: Use CDS server in France, default is CfA in US\n");
    fprintf (stderr,"-d: Print coordinates in degrees instead of sexigesimal\n");
    fprintf (stderr,"-e: Print coordinates in ecliptic instead of J2000\n");
    fprintf (stderr,"-g: Print coordinates in galactic instead of J2000\n");
    fprintf (stderr,"-i: Print ID(s) returned from SIMBAD\n");
    fprintf (stderr,"-n: Find coordinates using the NED database\n");
    fprintf (stderr,"-s: Find coordinates using SIMBAD\n");
    fprintf (stderr,"-t: Print output as tab-separated table\n");
    fprintf (stderr,"-v: Print extra descriptive info\n");
    fprintf (stderr,"-z: Find coordinates in Vizier catalog\n");
    exit (1);
}

/* Oct 25 2002	New program based on nedpos.c
 *
 * Jun 20 2006	Clean up code
 *
 * Jan 10 2007	Declare RevMsg static, not const
 * Jan 10 2007	exit(0) if successful
 * Jan 11 2007	Include fitsfile.h instead of fitshead.h
 * Jan 16 2007	Fix leading space bug
 * Sep 19 2007	Add -t option for Starbase output and @ for lists of names
 *
 * Apr 16 2008	Fix @ mode for processing lists of objects
 * 
 * Mar  3 2009	Put -i option back in; print name only if requested
 *
 * Jan 28 2013	Update Vizier URL
 * Feb  1 2013	Add -s NSV option to allow users to change search order
 * Feb  1 2013	Add -a option to print everything which is returned
 * Feb  1 2013	Fix -i option to print list of catalog IDs returned
 *
 * May  5 2015	Switch to CfA Vizier server as default
 * May 15 2015	Explicitly search SIMBAD, NED, or Vizier (nedpos or vizpos)
 */
